### PR TITLE
Add DEVICE_nGnRE to mmu_lpae 

### PR DESCRIPTION
**Pull request by jellesels-arm**
Add support for DEVICE_nGnRE to mmu_lpae. DEVICE_nGnRE memory is needed by the FF-A spec.
The SP device memory defined in the SP manifest file needs to be mapped as DEVICE_nGnRE.

This is a continuation from https://github.com/OP-TEE/optee_os/pull/5102/commits/c043bffa8e790cf0b00b5f39181386d58cb9ba43

### CODE REVIEW

**Review by jenswi-linaro — COMMENTED** [review]
Comments for "core: Add mattr_is_cached"

In the commit message:
`s/determen/determine/`

File: core/include/mm/tee_mmu_types.h [hunk 1]
```diff
@@ -34,8 +34,20 @@
...
+ * Device-nGnRnE most restrictive (equivalent to Strongly Ordered memory
+ * in the ARMv7 architecture).
+ * https://developer.arm.com/documentation/den0024/a/Memory-Ordering/Memory-types/Device-memory
+ *
+ * If an ARMv7 architecture operating system runs on a Cortex-A53 processor,
+ * the Device memory type matches the nGnRE encoding and the Strongly-Ordered
+ * memory type matches the nGnRnE memory type.
+ * https://developer.arm.com/documentation/den0024/a/Memory-Ordering/Memory-types/Device-memory
+ */
+#define TEE_MATTR_MEM_TYPE_DEV	        U(0) /* Device-nGnRnE */
```
Comment (jenswi-linaro) [inline core/include/mm/tee_mmu_types.h:48]: nGnRE?

File: core/arch/arm/mm/core_mmu_lpae.c [hunk 2]
```diff
@@ -329,7 +331,7 @@ static uint32_t desc_to_mattr(unsigned level, uint64_t desc)
 	if (desc & UPPER_ATTRS(PXN))
 		a &= ~TEE_MATTR_PX;
 
-	COMPILE_TIME_ASSERT(ATTR_DEVICE_INDEX == TEE_MATTR_MEM_TYPE_DEV);
+	COMPILE_TIME_ASSERT(ATTR_DEVICE_nGnRE_INDEX == TEE_MATTR_MEM_TYPE_DEV);
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu_lpae.c:334]: Please add a compile time assert for `TEE_MATTR_MEM_TYPE_STRONGLY_O` too.

Comment (jellesels-arm) [reply core/arch/arm/mm/core_mmu_lpae.c:336]: Ack

File: core/include/mm/tee_mmu_types.h (hunk 1 above)
Comment (jellesels-arm) [reply core/include/mm/tee_mmu_types.h]: Ack

File: core/include/mm/tee_mmu_types.h [hunk 3]
```diff
@@ -127,4 +127,11 @@ static inline void mattr_perm_to_str(char *str, size_t size, uint32_t attr)
 	str[6] = '\0';
 }
 
+static inline bool mattr_is_cached(uint32_t mattr)
+{
+	uint32_t mem_type = (mattr >> TEE_MATTR_MEM_TYPE_SHIFT) &
+		TEE_MATTR_MEM_TYPE_MASK;
```
[the remark this answers is not in the document]
Comment (jellesels-arm) [reply core/include/mm/tee_mmu_types.h]: Ack

File: core/include/mm/core_mmu.h [hunk 4]
```diff
@@ -534,8 +534,7 @@ enum teecore_memtypes core_mmu_get_type_by_pa(paddr_t pa);
 /* routines to retreive shared mem configuration */
 static inline bool core_mmu_is_shm_cached(void)
 {
-	return core_mmu_type_to_attr(MEM_AREA_NSEC_SHM) &
-		(TEE_MATTR_MEM_TYPE_CACHED << TEE_MATTR_MEM_TYPE_SHIFT);
+	return core_mmu_type_to_attr(MEM_AREA_NSEC_SHM);
```
[the remark this answers is not in the document]
Comment (jellesels-arm) [reply core/include/mm/core_mmu.h]: Ack

**Comment by jellesels-arm** [comment]
> 

I added a separate patch for it

**Comment by jenswi-linaro** [comment]
Please squash in the fixups.
I'd rather combine "core: Add support for DEVICE_nGnRnE in mmu_lpae" and "core: Add support for Strongly-ordered in mmu_v7" into one commit to avoid having partial support for the new memory type.

**Comment by jellesels-arm** [comment]
> 

Ack

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/arch/arm/mm/core_mmu_v7.c [hunk 5]
```diff
@@ -697,7 +698,7 @@ static void print_mmap_area(const struct tee_mmap_region *mm __maybe_unused,
 		debug_print("%s [%08" PRIxVA " %08" PRIxVA "] %s-%s-%s-%s",
 				str, mm->va, mm->va + mm->size,
 				mm->attr & (TEE_MATTR_CACHE_CACHED <<
-					TEE_MATTR_CACHE_SHIFT) ? "MEM" : "DEV",
+				TEE_MATTR_MEM_TYPE_SHIFT) ? "MEM" : "DEV"
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu_v7.c]: restore coma `,` at line end

File: core/arch/arm/mm/tee_pager.c [hunk 6]
```diff
@@ -566,8 +566,8 @@ static void *pager_add_alias_page(paddr_t pa)
 	struct core_mmu_table_info *ti;
 	/* Alias pages mapped without write permission: runtime will care */
 	uint32_t attr = TEE_MATTR_VALID_BLOCK |
-			(TEE_MATTR_CACHE_CACHED << TEE_MATTR_CACHE_SHIFT) |
-			TEE_MATTR_SECURE | TEE_MATTR_PR;
+			(TEE_MATTR_MEM_TYPE_CACHED << TEE_MATTR_MEM_TYPE_SHIFT)
+			| TEE_MATTR_SECURE | TEE_MATTR_PR;
```
Comment (etienne-lms) [inline core/arch/arm/mm/tee_pager.c:570]: for consistency in optee_os sources, `|` should be at the end of the previous line.
suggestion:
```c
	uint32_t attr = TEE_MATTR_VALID_BLOCK | TEE_MATTR_SECURE |
			TEE_MATTR_PR | (TEE_MATTR_MEM_TYPE_CACHED <<
					TEE_MATTR_MEM_TYPE_SHIFT);

**Review by etienne-lms — COMMENTED** [review]
Minor typos in commit logs:

> core: Chng TEE_MATTR_CACHE_ to TEE_MATTR_MEM_TYPE_
>
> Some extra memory types will be added. This patch ~~**makes**~~ renames all
TEE_MATTR_CACHE_ defines to TEE_MATTR_MEM_TYPE_. This will make the next
patches easier to understand.


> core: Add mattr_is_cache**d()**
> 
> mattr_is_cached() can be used to determ**~~en~~ine** if the mattr is cached or
not.

> core: Add support for DEVICE_nGnRnE
>
> Currently OP-TEE only allows non-cached memory to be mapped as
ATTR_DEVICE_nGnRE/Device. This patch adds support**~~ed~~** for
ATTR_DEVICE_nGnRnE/Strongly-ordered.

**Comment by jellesels-arm** [comment]
IBART gives the following errors:

**Comment by jenswi-linaro** [comment]
Try to rebase on master.

**Comment by jellesels-arm** [comment]
> Try to rebase on master.

Thanks, that worked!

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
Minor typos in commit messages.

> core: **~~Chng~~ change** TEE_MATTR_CACHE_ to TEE_MATTR_MEM_TYPE_
...

> core: Add mattr_is_cached()
> mattr_is_cached() can be used to **~~determen~~ determine** if the mattr is cached or
not.
> ...

**Comment by jellesels-arm** [comment]
> Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>
Ack

**Pull request merged**